### DISCUSSION TITLE
Speaking a date seems to result in not getting a final transcript

### DISCUSSION THREAD

**Discussion by davehorton** [deepgram/community#384] [General help]
### Which Deepgram product are you using?

Deepgram API

### Details

I am using the deepgram streaming api (websockets).  The path below indicates the model selection etc
```
/v1/listen?model=phonecall&tier=nova&language=en-US&smart_format=true&numerals=true&interim_results=true&endpointing=1500&encoding=linear16&sample_rate=8000
```
I have noticed that if I speak a date, I get a series of interim transcripts but it seems like I never get a final transcript.

For instance, I spoke: "My birthday is march 13, 1969".

I got a good interim transcript:

```
{"type":"Results","channel_index":[0,1],"duration":10.979937,"start":3.02,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"My birthday is 03/13/1969.","confidence":0.9989476,"words":[{"word":"my","start":3.34,"end":3.58,"confidence":0.5646759,"punctuated_word":"My"},{"word":"birthday","start":3.58,"end":4.08,"confidence":0.9989476,"punctuated_word":"birthday"},{"word":"is","start":4.14,"end":4.54,"confidence":0.9999225,"punctuated_word":"is"},{"word":"03/13/1969","start":4.54,"end":7.7599998,"confidence":0.9960316,"punctuated_word":"03/13/1969."}]}]},"metadata":{"request_id":"a75b0210-58fa-41c0-b5a7-2c62df33d4bb","model_info":{"name":"phonecall-nova","version":"2023-03-13.31000","arch":"nova"},"model_uuid":"47cb79bc-c75a-43c9-9cd7-90959e2e2b8c"}}
```
As you can see the transcript is accurate, but speech_final is false.  I stopped speaking and I continued to get interim transcripts at about 1 second intervals with the same result, but never got a final transcript.  Given that I set endpointing to 2000 I expected a final transcript 2 secs after finishing speaking

### If you are making a request to the Deepgram API, what is the full Deepgram URL you are making a request to?

/v1/listen?model=phonecall&tier=nova&language=en-US&smart_format=true&numerals=true&interim_results=true&endpointing=1500&encoding=linear16&sample_rate=8000

### If you are making a request to the Deepgram API and have a request ID, please paste it below:

a75b0210-58fa-41c0-b5a7-2c62df33d4bb

**Comment by jjmaldonis** [comment]
Hey @davehorton, this is an interesting edge case and the intended functionality comes down to how humans speak (particularly over the phone). When we speak numbers we often pause for much longer than usual between words.

If you couple these long pauses with the need to correctly format numbers, we run into a significant issue: should the formatter wait for the speaker to finish speaking the numbers in order to properly format them, or should the formatter work with the numbers it has and (probably) format them incorrectly? We've opted for the former because our customer research found that users tend to prefer correctly formatted numbers + long latency over incorrectly formatted numbers + short latency. So basically, accuracy is more important than low latency.

The issue you're seeing occurs when the "long pause between speaking numbers" is actually the end of the sentence. Since we ignore endpointing when a number is spoken, `speech_final` remains `false`.

We have a [docs page](https://developers.deepgram.com/docs/understand-endpointing-interim-results) that explains the interplay between interim results and endpointing, although it does not go into depth about the interaction I described above.

The typical solution is to add a new query parameter: `no_delay=true`. This query parameter is briefly documented [here](https://developers.deepgram.com/docs/smart-format#results). Adding this parameter may result in numbers that are poorly formatted if the speaker pauses for more than 2 seconds (your value for `endpointing`).

An alternative is to implement your own version of endpointing and manually set `speech_final` to `true` after N seconds of not receiving a message over the websocket. You could set N to 5 seconds for example, which should be more than enough time for the speaker to continue speaking if they had planned to, and will limit the latency to 5 seconds.

**Reply by nikolawhallon** [reply]
The formatting issue is definitely happening, but there is another issue when setting `endpointing` to a large value, say longer than 1 second - the `endpointing` algorithm is audio-based, so within those 2 seconds if there is any blip of audio, not necessarily speech, that would throw the `endpointing` algorithm off, and our other chunking algorithms may make finalized chunks instead (assuming the formatter is not waiting for more words, i.e. if you use `no_delay`). If this becomes a problem we have another option to try - `utterance_end_ms=2000` which is a word-based endpointing algorithm which looks to see "have any words been spoken in 2000 ms?" and if not, it will send a message of the form `{"type":"UtteranceEnd"}`. This is robust against background noise, but requires that our model actually process the audio to create transcripts - the limitation there is that it has a resolution of only about 1 second (so values of 1000 ms, 2000 ms, 3000 ms, etc, are fine, but a value of 400 ms wouldn't do any good).

**Comment by davehorton** [comment]
just so I understand, `utterance_end_ms` and `endpointing` are mutually exclusive, correct?  I can choose one, or the other (or neither) but I should not choose to use both on the same command?

**Reply by nikolawhallon** [reply]
It's ok to use both at the same time, since `utterance_end_ms` results in independent json messages being sent, while `endpointing` makes its indication with the `speech_final` field in a regular ASR json response, but in your application logic you should pick which of these to trigger off of (my apps use `utterance_end_ms` typically).

**Comment by davehorton** [comment]
Thanks for the detailed help on this.  I did try the no_delay param:

```
/v1/listen?model=phonecall&tier=nova&language=en-US&smart_format=true&no_delay=true&numerals=true&interim_results=true&endpointing=600&encoding=linear16&sample_rate=8000
```
and it seemed to resolve my issue, as you had indicated.  I now get a final transcript:
```
{
    "type": "Results",
    "channel_index": [
        0,
        1
    ],
    "duration": 4.5299997,
    "start": 3.97,
    "is_final": true,
    "speech_final": true,
    "channel": {
        "alternatives": [
            {
                "transcript": "My birthday is 03/16/1964.",
                "confidence": 0.94970703,
                "words": [
                    {
                        "word": "my",
                        "start": 3.97,
                        "end": 4.21,
                        "confidence": 0.7421875,
                        "punctuated_word": "My"
                    },
                    {
                        "word": "birthday",
                        "start": 4.21,
                        "end": 4.69,
                        "confidence": 0.94970703,
                        "punctuated_word": "birthday"
                    },
                    {
                        "word": "is",
                        "start": 4.69,
                        "end": 4.93,
                        "confidence": 0.9902344,
                        "punctuated_word": "is"
                    },
                    {
                        "word": "03/16/1964",
                        "start": 4.93,
                        "end": 7.67,
                        "confidence": 0.84052736,
                        "punctuated_word": "03/16/1964."
                    }
                ]
            }
        ]
    },
    "metadata": {
        "request_id": "4c917df7-51e0-469a-90a1-656434660bd2",
        "model_info": {
            "name": "phonecall-nova",
            "version": "2023-03-13.31000",
            "arch": "nova"
        },
        "model_uuid": "47cb79bc-c75a-43c9-9cd7-90959e2e2b8c"
    }
}
```

The `utterance_end_ms` is something I will look at next as it does seem useful in certain use cases (like entering a long string of digits, like a credit card number.

One final question.  I notice that when I start transcribing I immediately am getting a transcript with `is_final: true` and `speech_final: true`, but the transcript is empty:
```
{"type":"Results","channel_index":[0,1],"duration":1.25,"start":0.0,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"4c917df7-51e0-469a-90a1-656434660bd2","model_info":{"name":"phonecall-nova","version":"2023-03-13.31000","arch":"nova"},"model_uuid":"47cb79bc-c75a-43c9-9cd7-90959e2e2b8c"}}
```

What is the meaning of this?  I have been ignoring them for now, is that a reasonable thing to do?

Thanks again for your insights

**Reply by nikolawhallon** [reply]
Since `endpointing` is audio-based, it's possible some noise at the beginning of the audio trigger it to think there was speech, and when the audio got quiet again, it triggered (it also fires in instances when you turn on a fan, for example, or start playing music, or a ringing phone starts and then stops). Using some heuristic where all `is_final` results are concatenated until the first `speech_final`, and making sure the transcript makes sense (e.g. is not empty) is very reasonable.

**Comment by davehorton** [comment]
I'm trying to implement the suggested heuristic, but having some issues
```
Using some heuristic where all is_final results are concatenated until the first speech_final, and making sure the transcript makes sense (e.g. is not empty) is very reasonable.
```

In this case, I set endpointing to 1200 milliseconds:
```
/v1/listen?model=phonecall&tier=nova&language=en-US&smart_format=true&no_delay=true&numerals=true&interim_results=true&endpointing=1200&encoding=linear16&sample_rate=8000
```

I then said: "I think..{brief pause}...that I could make the meeting..{brief pause}..tomorrow..{long pause}"

My expectation was that I would get some 3 transcripts back with `is_final` but not `speech_final` for the 3 chunks of audio, and then after a delay corresponding to the 1200 milliseconds I would get a final transcript (maybe empty) with `speech_final`.

However that is not what I experienced.  I did get the 3 expected transcripts with `is_final`:
```
 {"type":"Results","channel_index":[0,1],"duration":3.0500002,"start":3.02,"is_final":true,"speech_final":false,"channel":{"alternatives":[{"transcript":"I think.","confidence":0.77504575,"words":[{"word":"i","start":3.34,"end":3.58,"confidence":0.6579492,"punctuated_word":"I"},{"word":"think","start":3.58,"end":4.08,"confidence":0.77504575,"punctuated_word":"think."}]}]},"metadata":{"request_id":"446f8fdf-2157-48f8-9b99-c3e509879db0","model_info":{"name":"phonecall-nova","version":"2023-03-13.31000","arch":"nova"},"model_uuid":"47cb79bc-c75a-43c9-9cd7-90959e2e2b8c"}}

{"type":"Results","channel_index":[0,1],"duration":4.5799994,"start":6.07,"is_final":true,"speech_final":false,"channel":{"alternatives":[{"transcript":"That I could make the meaning","confidence":0.9937494,"words":[{"word":"that","start":6.55,"end":7.05,"confidence":0.98417413,"punctuated_word":"That"},{"word":"i","start":8.95,"end":9.190001,"confidence":0.9937494,"punctuated_word":"I"},{"word":"could","start":9.190001,"end":9.43,"confidence":0.99518216,"punctuated_word":"could"},{"word":"make","start":9.43,"end":9.75,"confidence":0.9989098,"punctuated_word":"make"},{"word":"the","start":9.75,"end":9.83,"confidence":0.86287236,"punctuated_word":"the"},{"word":"meaning","start":9.83,"end":10.33,"confidence":0.67458856,"punctuated_word":"meaning"}]}]},"metadata":{"request_id":"446f8fdf-2157-48f8-9b99-c3e509879db0","model_info":{"name":"phonecall-nova","version":"2023-03-13.31000","arch":"nova"},"model_uuid":"47cb79bc-c75a-43c9-9cd7-90959e2e2b8c"}}

 {"type":"Results","channel_index":[0,1],"duration":4.67,"start":10.65,"is_final":true,"speech_final":false,"channel":{"alternatives":[{"transcript":"tomorrow.","confidence":0.97822493,"words":[{"word":"tomorrow","start":10.65,"end":11.15,"confidence":0.97822493,"punctuated_word":"tomorrow."}]}]},"metadata":{"request_id":"446f8fdf-2157-48f8-9b99-c3e509879db0","model_info":{"name":"phonecall-nova","version":"2023-03-13.31000","arch":"nova"},"model_uuid":"47cb79bc-c75a-43c9-9cd7-90959e2e2b8c"}}
```
But after that I never got a `speech_final`.

Again, I just continued to get empty transcripts with is_final=false and speech_final=false for many seconds until I hung up

```
{"type":"Results","channel_index":[0,1],"duration":2.9599361,"start":19.12,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"446f8fdf-2157-48f8-9b99-c3e509879db0","model_info":{"name":"phonecall-nova","version":"2023-03-13.31000","arch":"nova"},"model_uuid":"47cb79bc-c75a-43c9-9cd7-90959e2e2b8c"}}
```

The request id was `446f8fdf-2157-48f8-9b99-c3e509879db0` if that helps.

Is this expected behavior?  When collecting an audio input that I expect to have pauses I was counting on being able to set the endpointing property to help me be able to collect this.  I believed that I was guaranteed to eventually get a `speech_final` result but per the above that is not the case.

(I understand you have the `utterance_end_ms` per the discussion above, and I am doing some testing with that) but I don't want to have to rely on that for every use case).

In the example above I did set the `no_delay` property

**Comment by davehorton** [comment]
to summarize the responses I got:
```
time               is_final   speech_final   transcript
-------------------------------------------------------
14:00:51.742657    no         no              ""
14:00:52.762657    no         no              ""
14:00:53.622658    no         no              ""
14:00:54.642657    no         no              "I think."
14:00:55.542658    no         no              "I think"
14:00:55.742660    yes        no              ""
14:00:55.762660    no         no              "I think"
14:00:56.022657    no         no              "I think."
14:00:56.282658    no         no              "I think."
14:00:57.082658    no         no              "I think."
14:00:58.082657    no         no              "I think that"
14:00:58.622658    no         no              "I think that"
14:00:58.662665    yes        no              "I think."
14:00:58.682664    no         no              "That"
14:00:58.862658    no         no              "That"
14:00:59.362657    no         no              "That"
14:00:59.442659    no         no              "That"
14:00:59.662661    no         no              "That"
14:01:00.822658    no         no              "That I could make the"
14:01:01.422658    no         no              "That I could make the meaning"
14:01:01.802658    yes        no              "That I could make the meaning"
14:01:02.822658    no         no              "tomorrow."
14:01:02.962658    no         no              "tomorrow."
14:01:03.082658    no         no              "tomorrow."
14:01:03.462657    no         no              "tomorrow."
14:01:03.642658    no         no              "tomorrow."
14:01:03.782657    no         no              "tomorrow."
14:01:04.102658    no         no              "tomorrow."
14:01:04.582657    no         no              "tomorrow."
14:01:04.902658    no         no              "tomorrow."
14:01:05.182657    no         no              "tomorrow."
14:01:05.802657    no         no              "tomorrow."
14:01:06.202657    no         no              "tomorrow."
14:01:06.402659    yes        no              "tomorrow."
14:01:07.042656    no         no              ""
14:01:07.242659    no         no              ""
..repeats for many seconds..
14:01:12.842658    no         no              ""
..then I hang up
```

Shouldn't I be guaranteed to receive a `speech_final=true` after (in this case) 1200 millseconds of silence?  Otherwise, the rule/heuristic you provided above for compiling reponses does not work

**Reply by nikolawhallon** [reply]
A couple things here make sense:
* if you concatenate all `is_final` results, you do get the full transcript
* a `speech_final` result never occurs, but this could just be because in the 1.2 seconds after "tomorrow" was said, any blip of audio could have made the endpointing algorithm fail to trigger (it's best for short pauses to make minimum latency, the default is 10 ms)
* a `speech_final` result is not guaranteed after 1200 ms of silence, only after 1200 ms of silence with preceding speech audio, so if a new chunk starts forming at `14:01:07.042656` and there is never any speech after that, the state will never have gone from speech -> no speech, and `speech_final`/endpointing will never trigger

It looks like from `14:01:07.242659` to `14:01:12.842658` there are no `is_final` results, this is the only problem that I can't explain... we /do/ guarantee an `is_final` result at least every 3-5 seconds (this can be configured for onprem customers). Also, if the connection to Deepgram is closed gracefully, we do flush out any remaining results and force the last transcription result to be a final/`is_final` result.

I can look into that request id to see why that didn't seem to happen tomorrow! Thanks for finding this interesting case.
